File: app/src/main/java/com/hsilighting/pagify/core/RenderScale.kt
package com.hsilighting.pagify.core

import kotlin.math.ceil
import kotlin.math.sqrt

/**
 * The single definition of "how many pixels per PDF point do we rasterise at".
 *
 * This lives on its own because two callers need the *identical* answer: the page
 * view that renders what you see, and the prefetcher that warms the native cache
 * ahead of you. They previously computed it separately — and disagreed, one
 * passing a points-to-pixels scale and the other a fit-width magnification factor.
 * Every prefetch therefore landed under a cache key the renderer never looked up,
 * so the cache never once produced a hit in normal use. Sharing this function is
 * what makes that class of bug impossible rather than merely fixed.
 *
 * Note the distinction the names now keep straight:
 *  - **zoom** is user-facing magnification, 1.0 == fit to viewport width.
 *  - **render scale** is pixels per point, which is what the engine wants.
 */
object RenderScale {

    /** Must match `ZOOM_QUANTUM` in `rust/pdf_core/src/render/cache.rs`. */
    const val QUANTUM = 0.25f

    /**
     * Per-page bitmap ceiling: 16 MP, i.e. 64 MB at 4 bytes per pixel.
     *
     * A page is a single bitmap, so its cost grows with the square of zoom — an A4
     * page at 8x would be ~930 MB and would fail in `Bitmap.createBitmap` well
     * before the engine's own size guard could reject it. Past this ceiling the
     * bitmap is upscaled for display instead: softer at extreme zoom, but it
     * cannot bring the app down.
     */
    const val MAX_PIXELS = 16_000_000.0

    /**
     * Render scale for [pageSize] drawn [targetPixelWidth] pixels wide.
     *
     * Quantised *up* to [QUANTUM] steps to match the native cache's own
     * quantisation: without that, every pixel of pinch drift would be a fresh
     * cache key, and rounding down would leave the raster slightly too blurry.
     */
    /**
     * Fraction of the readable width a proxy render uses.
     *
     * A quarter of the width is a sixteenth of the pixels, which is the
     * difference between roughly 5 ms and roughly 80 ms per page. Recognisable
     * while scrolling, and never mistaken for the readable pass.
     */
    const val PROXY_FRACTION = 0.25f

    /** Width of a navigator/strip thumbnail, in pixels. */
    const val THUMBNAIL_WIDTH_PX = 190f

    /** The cheap first pass. See [PROXY_FRACTION]. */
    fun proxyFor(pageSize: PageSize, targetPixelWidth: Float): Float =
        forPage(pageSize, targetPixelWidth * PROXY_FRACTION)

    /**
     * Scale for a thumbnail, independent of zoom.
     *
     * Fixed width rather than a fraction of the reading size so that thumbnails
     * keep one cache key no matter how far the reader is zoomed in — otherwise
     * every zoom step would orphan the entire strip.
     */
    fun thumbnailFor(pageSize: PageSize): Float = forPage(pageSize, THUMBNAIL_WIDTH_PX)

    fun forPage(pageSize: PageSize, targetPixelWidth: Float): Float {
        if (pageSize.widthPoints <= 0f || pageSize.heightPoints <= 0f) return QUANTUM
        if (!targetPixelWidth.isFinite() || targetPixelWidth <= 0f) return QUANTUM

        val ideal = targetPixelWidth / pageSize.widthPoints
        val maxScale = sqrt(
            MAX_PIXELS / (pageSize.widthPoints.toDouble() * pageSize.heightPoints.toDouble()),
        ).toFloat()

        val bounded = ideal.coerceAtMost(maxScale)
        return (ceil((bounded / QUANTUM).toDouble()) * QUANTUM).toFloat().coerceAtLeast(QUANTUM)
    }
}

File: ios/Sources/RenderScale.swift
import CoreGraphics
import Foundation

/// Choosing what scale to rasterise a page at.
///
/// Every render goes through here, for two reasons the app cannot see and the
/// engine cannot fix on its own.
///
/// **The cache is keyed on a quantised zoom.** A pinch produces a continuum of
/// float scales, and asking for 2.0413 then 2.0417 fills the engine's LRU with
/// near-duplicates while missing on every frame. Quantising here — to the same
/// step the engine uses — is what makes the cache a cache.
///
/// **Nothing else caps the size.** A full page at an unbounded zoom is an
/// arbitrarily large allocation, and the engine's own ceiling refuses it rather
/// than trimming it, so the page simply fails to draw at the moment the reader
/// zoomed in to look at it.
enum RenderScale {
    /// Must equal `ZOOM_QUANTUM` in `rust/pdf_core/src/render/cache.rs`. If the
    /// two ever disagree the cache silently stops hitting.
    static let quantum: CGFloat = 0.25

    /// The most pixels worth rasterising for one page.
    static let maxPixels: CGFloat = 16_000_000

    /// The first, cheap pass while a zoom is still moving. Measured on Android at
    /// about 5 ms against 50–99 ms for the readable pass.
    static let proxyFraction: CGFloat = 0.25

    /// What the rail asks for.
    static let thumbnailWidthPx: CGFloat = 190

    /// The scale to render `pageSize` at to fill `targetPixelWidth`.
    static func forPage(_ pageSize: CGSize, targetPixelWidth: CGFloat) -> CGFloat {
        guard pageSize.width > 0, pageSize.height > 0, targetPixelWidth > 0 else {
            return quantum
        }

        let ideal = targetPixelWidth / pageSize.width

        // The area ceiling, expressed as the largest scale that still fits.
        let maxScale = sqrt(maxPixels / (pageSize.width * pageSize.height))

        // Rounded *up*, so the raster is never lower-resolution than asked for —
        // scaling a bitmap down is free and sharp, scaling up is neither.
        var stepped = (min(ideal, maxScale) / quantum).rounded(.up) * quantum

        // Rounding up can push it back over the ceiling; step down once if so.
        if stepped > maxScale { stepped -= quantum }

        return max(stepped, quantum)
    }

    /// The scale a thumbnail wants.
    ///
    /// A rail thumbnail is a fixed pixel width whatever the page's shape, so it
    /// goes through the same quantiser as everything else — a rail of ten pages
    /// asking ten unquantised scales is ten cache misses on every scroll.
    static func thumbnailFor(_ pageSize: CGSize,
                             targetPixelWidth: CGFloat = thumbnailWidthPx) -> CGFloat {
        forPage(pageSize, targetPixelWidth: targetPixelWidth)
    }

    /// The cheap pass, for while the zoom is still moving.
    static func proxyFor(_ pageSize: CGSize, targetPixelWidth: CGFloat) -> CGFloat {
        forPage(pageSize, targetPixelWidth: targetPixelWidth * proxyFraction)
    }
}

/// How far the reader can zoom, and what the gestures snap to.
enum Zoom {
    static let minimum: CGFloat = 1
    static let maximum: CGFloat = 8
    /// The page fits the width of the reader.
    static let fitWidth: CGFloat = 1
    static let doubleTap: CGFloat = 2.5
    /// How much of a pinch it takes, in the page list, before the reader is
    /// handed over to the zoomed view.
    static let pinchHandover: CGFloat = 1.12

    static func clamp(_ scale: CGFloat) -> CGFloat {
        min(max(scale, minimum), maximum)
    }
}


/// Admits a bounded number of full-page rasters at once.
///
/// A recording of a scroll through a 149-page catalogue caught **four** page
/// renders finishing on the same millisecond, 280ms each: four PDFium rasters
/// competing for the same cores, and the scroll stuttering behind them. Over half
/// the stalls in that session had a render landing inside them.
///
/// Nothing here was bounded — each row started its own `Task.detached`. Android
/// does not have the problem in the same way: its page renders go to
/// `Dispatchers.Default`, which is already bounded by core count, and its
/// thumbnails hold a `Semaphore(1)`. iOS matched Android on thumbnails —
/// `ThumbnailRenderer` is an actor, so one at a time — and matched nothing at all
/// on pages.
///
/// Two, not one: the page being read and the one arriving behind it can both be
/// working without leaving the scroll to fight them for a core.
actor RenderGate {
    static let shared = RenderGate(limit: 2)

    private let limit: Int
    private var running = 0
    private var waiting: [CheckedContinuation<Void, Never>] = []

    init(limit: Int) { self.limit = limit }

    /// Run `body` once there is room, and make room again afterwards.
    nonisolated func run<T>(_ body: () async -> T) async -> T {
        await enter()
        let result = await body()
        await leave()
        return result
    }

    private func enter() async {
        if running < limit {
            running += 1
            return
        }
        await withCheckedContinuation { waiting.append($0) }
    }

    private func leave() {
        // Handed straight to whoever is next, so the count never dips and lets a
        // third render in behind them.
        if waiting.isEmpty {
            running -= 1
        } else {
            waiting.removeFirst().resume()
        }
    }
}
